### DISCUSSION TITLE
How do I iterate over a list?

### DISCUSSION THREAD

**Discussion by moncheeta** [nlohmann/json#3387] [Q&A] [answered]
I have a program where I have to access the objects in a list of a json file. This is not an exact copy, but it looks something like this:
```json
{
  "Objects": [
    {
       "Name": "Blah",
       "...": "everything else"
    }
  ]
}
```
There can be a random amount of objects in the list. So I tried something like this:
```cpp
int main() {
  json data;
  std::ifstream file("./objects.json", std::ifstream::in);
  file >> data;
  std::vector<object> all_objects;
  for (auto &object_iter : file) {
    all_obejcts.push_back(object(object_iter["Name"], object_iter["..."]));
  }
  file.close();
  std::cout << "Use the data..." << std::endl;
}
```
I get an error: ``what():  [json.exception.type_error.302] type must be number, but is string``

**Comment by gregmarr** [comment, marked as answer]
That code is invalid.  You're iterating over the `ifstream` and not the `nlohmann::json`, and you have an error in the vector name inside the `for` loop.  Do you have actual code that you can share?  If your code is otherwise correct, and your JSON sample is correct, then in the first iteration of the loop, `object_iter` is pointing at the json object with a name of `"Objects"` and a value of an array.

Is `object` in `object(object_iter["Name"], object_iter["..."])` `json::object` ?

**Reply by moncheeta** [reply]
Original Code(It takes all the pokemon json files in ./pokemon):
```cpp
std::vector<pokemon> get_all_pokemon() {
  std::vector<pokemon> all_pokemon;
  for (const std::experimental::filesystem::directory_entry &file : std::experimental::filesystem::directory_iterator("./pokemon")) {
    json data;
    std::ifstream pokemon_file(file.path(), std::ifstream::in);
    pokemon_file >> data;
    std::vector<attack> new_pokemon_attacks;
    for (auto &attackinjson : data["Attacks"]) {
      attack new_attack = { attackinjson["Name"], attackinjson["Damage"] };
      new_pokemon_attacks.push_back(new_attack);
    }
    pokemon new_pokemon(data["Name"], 0, data["Health"], data["Character"], new_pokemon_attacks);
    all_pokemon.push_back(new_pokemon);
    pokemon_file.close();
  }
  return all_pokemon;
}
```
"pickachu.json" file in "./pokemon":
```json
{
	"Name": "Pickachu",
	"Health": "100",
	"Character": "                   ,              \n              :K              \n              0O              \n              0l              \n              K'              \n              0xkxl'   .;:'   \n              0KKKKKddKKKO.   \n      .;d;   :KKKKKKKk        \n    ,xKKK.   kdkKKKKKO        \n  c0KKK0Ooo  K';KKK0OO        \n.0KKKKOx:kKk,:00KO0l;c        \n.KKK0kxx.cKKc;O0,,xOo,        \n KK0k;;x .KKddK0;;'0O.        \n dKk  ;x .0K0OKKclxKllKK,     \n .k   od;dlKK0KKodK0d0K0      \n      xxd: OKK0kOO0OKKO       \n      , ,. OKKK0O0K0O.        \n        ..cKKKKKKKK:          \n          OKKKKKKKK.          \n         :KKKKKKKKKc          \n         0KKKKKKKKK0.         \n        .00KKKKKKKKKx         \n         xxkOOkkxkk0o         \n         ooo.   dxxx'         \n        :          .l.        \n",
	"Attacks": [
		{
			"Name": "Lightning",
			"Damage": "50"
		}
	]
}
```

**Reply by gregmarr** [reply]
I don't see anything wrong with your code or your JSON.  Are you sure that's the JSON where it's failing?

[Compiler Explorer](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGe1wAyeAyYAHI%2BAEaYxBIArKQADqgKhE4MHt6%2BekkpjgJBIeEsUTFc8XaYDmlCBEzEBBk%2Bfly2mPZ5DDV1BAVhkdFxtrX1jVktCsM9wX3FA2UAlLaoXsTI7BzmAMzByN5YANQmm24MtKgILIYMAPTaCgIAdAgJCUfYJhoAgls7e5iHxycE2ImFYbw%2B3y%2BaAYE32E3QIBAwOCwH2NHoAIAIvsAErmMwQEwAdisXw%2BAE58aFWJh8SBDmYzMo8MgANZMZAILz40gQin4gASoNoBAQdIZZi4Gg0PL5GkpjLcCDqHII0XF%2BP2Wu1Ot1OtIesNh1iJyNepAAGkzbqTCaGNadRoAPIO7W202u/YaWie40e10WsAcX3u%2B2ejSqVmqWhB7UPI4WcDBrWh30aC0ZjPodCZp0PN12vXxzYWdAJ7WWzOZ1lG1Pa0iqTP5rWs9CsqsWl2GuvIdMZ52oVBawMbEuZ51dvWhh59jNO9Tt1mkEDSzvehPIWt2h7jqOqB7IDOmEtOjSkBtO1AG7uF/ushMJ1T7HcW7MWmUlhNB53NqeFnM1gCFhPjO75OketCNrQtAZteBamg8gH7BgCboDB74Zhg77oOmGh/v62qqKoCL7OB76sk6p5kZOOp1vq%2BykPmZFgbheZbgROoPAeHaZvSZp0YazEdr%2BN4cTqlY8Ram78YWRqzjxGgiTaslFqukkWk%2B7FhoaREUU6rJRgZGhDlpZqDqgv7oERqixqZhp8UaDy0Ep8EMLKZLyvinwELUbIKBqsSkpClKeSScohRFVI0hqjJBMACAEAwKLucF/KeYymKsEwwC0oy9L4rEMqMuF/JEpiJWxOVZJlfM%2BIPh5nzBAQ%2ByXMEEDzIcYVfDqTBeEQ%2BzaFi%2BynOclwMH4IB3M4IAJHUSjUHQmC1Z%2B3XavwxD7BAvX9eYABsTA%2BRyrLBFN9r0totpWIy3m%2Bay/nFZVHXEkFerbUOzBsENB23Sd9xuYFUVsHVlX1Z8r19UO6BZTlX2HWyv0CJd%2BKZZcOXA%2BVK1g7q8KImgfUAm4RxuMNNIE0TEogJqRPk1DqP/NTxwSqGdWY7RZUQjqIIECs9ofkFxKYhwiy0JwsS8H4HBaKQqCcITljWHCyyrPTZibDwpAEJoQuLKyIC7VwDxmBokjklI5JEpIXBEmY5LxCLHCSOLWvS5wvAKCuGta4scCwEgaAsAki1kBQED%2B4H9AxLshjAAA%2BgQxBeAwS7omqxDuxAETOxEwR1AAns7/tsIITqnPnku8FgY3AOI5ekPgIJVAAbpg7u15gqiVH16xS01bTO7QeARMQeceFgzvx3gLCcDwixUAYwAKAAangmAAO5OgkjDT7w/CCCIYjsFIMiCIoKjqLXugtAYRgoNY1j6IP7uwB97AjRcVwx6dpDNzE8eJ6y8xFioASB0Vubs2iVA6C4Bg7hPBNH8DA3oRQSjZGSKkAQoxmiJDQR0JB/RSitHaNUSYmC9AVCqAILo9Q8GzAIRMbopDxiTBoSgrgix7grDWHoeOmB1gz30KLJ2tcZYcFUAADl2gAWl2pIfYUcjCbT/knDqEA5ZXRsPsXAhASAMjVgaDwAcg66LYbwTW5dvZIGWAQBIfVyCUDDkHfw%2BAiAEN3sIUQ4gj5uNPmoceII%2BG8FXsPBI29haCNIBLKWIinR9Rsc1VAVB9jiKkTIuR19UQQCUQAzaBjw7RGMfMUxXtFgIFBFgGI7VSC602GIh4RJbYaDEYVMwsQxFmCJBoO2nBHYROdiIt23DikCI4GYIRUTXae3MYsH%2BKRnCSCAA%3D%3D)

**Reply by gregmarr** [reply]
I would suggest printing out the filename between these two lines

    std::ifstream pokemon_file(file.path(), std::ifstream::in);
    pokemon_file >> data;

and then once you know which file is failing, run it just on that file and do a `data.dump();`.

**Reply by moncheeta** [reply]
Sorry for the late response. ``data.dump()`` returns ``null``.

**Reply by moncheeta** [reply]
I don't know why it returns null as the correct path is being used.

**Reply by gregmarr** [reply]
Just to be sure, the `data.dump();` is after `pokemon_file >> data;`?

**Reply by gregmarr** [reply]
I guess the next step would be to read the file into a `std::string` so you can print the string, and then do `data = nlohmann::json::parse(str);`

**Reply by moncheeta** [reply]
> Just to be sure, the `data.dump();` is after `pokemon_file >> data;`?

I fixed it and I didn't get null. The output seemed perfect.

> I guess the next step would be to read the file into a `std::string` so you can print the string, and then do `data = nlohmann::json::parse(str);`

I got:
```
terminate called after throwing an instance of 'nlohmann::detail::parse_error'
  what():  [json.exception.parse_error.101] parse error at line 1, column 2: syntax error while parsing object key - unexpected end of input; expected string literal
Aborted (core dumped)
```

**Reply by gregmarr** [reply]
> I fixed it and I didn't get null. The output seemed perfect.

I don't understand this.  Does this mean that you had the `data.dump()` before the `pokemon_file >> data;` and once you put it after that, the output was correct?

> parse error at line 1, column 2: syntax error while parsing object key - unexpected end of input

Sounds like it's only getting `{` as its input here.

**Reply by gregmarr** [reply]
To be clear, the "read the file into a std::string so you can print the string, and then do data = nlohmann::json::parse(str);" was instead of the `pokemon_file >> data;`.

**Reply by moncheeta** [reply]
I by accident put ``data.dump()`` before giving data the contents of the file. When I put it in the correct place, I get the expected output.
The output of ``data.dump()``:
```json
{"Attacks":[{"Damage":"75","Name":"Fire"}],"Character":"           .O                 \n           xxl;               \n        l  odddd'.     l      \n     ;:l.   xxxd .     .      \n    :cc;'     dx        ;.    \n  ;clll;c     ,x'      ,':;   \n .cllll:c.    .xo     .,,:c   \n :lllll:c:.   .kx     ;:;clc. \n.c:  .l:l:,,. .Ox   .;,l;:llc \n;.    ;:cc,';;:Ok;;,,':l,;lll.\n,      ,  .:lkk0Oxo;; ,c,  lc.\n         .. ckOOOkx. l ,  :..,\n        lk  ckKNNKl  'l.  ;o ,\n        , ..O0NNNNO  ;c.  kx .\n           xONNNNN0.      KO. \n          dO0NNNNK0o      KX' \n     'llcdOOOXNNX00k:     o:  \n    ;ddod0OOx0KK000kOc    .   \n   ,xxdod0OxO00000OkOO' .'    \n   .oolldxdx00000xoxOkoo      \n     dddxxc        .xx'       \n       ;xd.          dl       \n        .                     ","Health":"120","Name":"Charzard"}
```
charzard.json(I forgot to include this):
```json
{
	"Name": "Charzard",
	"Health": "120",
	"Character": "           .O                 \n           xxl;               \n        l  odddd'.     l      \n     ;:l.   xxxd .     .      \n    :cc;'     dx        ;.    \n  ;clll;c     ,x'      ,':;   \n .cllll:c.    .xo     .,,:c   \n :lllll:c:.   .kx     ;:;clc. \n.c:  .l:l:,,. .Ox   .;,l;:llc \n;.    ;:cc,';;:Ok;;,,':l,;lll.\n,      ,  .:lkk0Oxo;; ,c,  lc.\n         .. ckOOOkx. l ,  :..,\n        lk  ckKNNKl  'l.  ;o ,\n        , ..O0NNNNO  ;c.  kx .\n           xONNNNN0.      KO. \n          dO0NNNNK0o      KX' \n     'llcdOOOXNNX00k:     o:  \n    ;ddod0OOx0KK000kOc    .   \n   ,xxdod0OxO00000OkOO' .'    \n   .oolldxdx00000xoxOkoo      \n     dddxxc        .xx'       \n       ;xd.          dl       \n        .                     ",
	"Attacks": [
		{
			"Name": "Fire",
			"Damage": "75"
		}
	]

```
The code that instead uses strings:
```cpp
std::vector<pokemon> get_all_pokemon() {
  std::vector<pokemon> all_pokemon;
  for (const std::experimental::filesystem::directory_entry &file : std::experimental::filesystem::directory_iterator("./pokemon")) {
    std::ifstream pokemon_file(file.path(), std::ifstream::in);
    std::string strfile;
    pokemon_file >> strfile;
    std::cout << strfile << std::endl;
    std::cin.get();
    json data = json::parse(strfile);
    pokemon_file.close();
    std::vector<attack> new_pokemon_attacks;
    for (auto &attackinjson : data["Attacks"]) {
      attack new_attack = { attackinjson["Name"], attackinjson["Damage"] };
      new_pokemon_attacks.push_back(new_attack);
    }
    pokemon new_pokemon(data["Name"], 0, data["Health"], data["Character"], new_pokemon_attacks);
    all_pokemon.push_back(new_pokemon);
  }
  return all_pokemon;
}
```
The code that uses strings only prints "{" when it gets printed(``std::cout << strfile << std::endl;``).

**Reply by gregmarr** [reply]
> The code that uses strings only prints "{" when it gets printed(std::cout << strfile << std::endl;).

Since parsing from the `ifstream` and doing `data.dump();` provides the correct information, I don't think you need to pursue this bit.

**Reply by gregmarr** [reply]
Oh, can you show the definition of the `pokemon` and `attack` classes?  There are only two places that result in that error message, and they're in getting a number from the `json` object.  My guess is that either attack damage or the pokemon health is a number, but you're trying to get that number from a string value in the JSON.

**Reply by moncheeta** [reply]
Sorry, I'm new to json and I though that you can only have strings. I replaced the appropriate variables to numbers and it works perfectly!

**Reply by gregmarr** [reply]
Great, thanks for the update.
